### PR TITLE
Add `BlockMetadata` to consensus, ledger and watcher

### PR DESCRIPTION
**Pull request by remoun**
Builds on #2058.

### Motivation

* #1403
* https://github.com/mobilecoinfoundation/mcips/pull/43

### Future Work
* #2104

### CODE REVIEW

File: transaction/core/src/blockchain/quorum_set.rs
Comment (remoun) [inline transaction/core/src/blockchain/quorum_set.rs:20]: I opted to define a minimal/simple `QuorumSet` after grappling with `scp::QuorumSet`, trying to make it implement `Message`, which `LedgerDB` requires to serialize values. Note that this PR also includes `impl [Try]From` to convert back and forth between them.

File: ledger/db/src/ledger_db.rs
Comment (remoun) [inline ledger/db/src/ledger_db.rs:567]: This is a key part of this PR

File: consensus/service/src/byzantine_ledger/worker.rs
Comment (remoun) [inline consensus/service/src/byzantine_ledger/worker.rs:886]: This is a key part of this PR

File: api/src/convert/archive_block.rs
Comment (iamalwaysuncomfortable) [inline api/src/convert/archive_block.rs:59]: If it's not already in one of the try statements above, the block id should be computed & matched. 

```suggestion
        if block.contents_hash == block_contents.hash() && block.is_block_id_valid() {
            Ok(BlockData::new(block, block_contents, signature, metadata))
```

**Review by iamalwaysuncomfortable — COMMENTED** [review]
Left an initial comment, will look with fresh eyes in the morning

File: consensus/service/src/byzantine_ledger/worker.rs
Comment (iamalwaysuncomfortable) [inline consensus/service/src/byzantine_ledger/worker.rs:884]: If we're signing the Block Metadata, this is definitely the unique message signing key that all consensus nodes are required to have (and send other peers to be considered part of the Quorum process). So if we're saying we're putting Block Metadata on chain signed by consensus nodes, I think this is definitely the key we want.

A subtle, yet significant TODO in this process is finding a mechanism to track a record of these signing keys. If we are forgetting the public key history of the consensus notes, it somewhat defeats the purpose of signing the block as we have no way to verify that these were the authoritative keys at a later point.

File: consensus/service/src/byzantine_ledger/worker.rs
Comment (iamalwaysuncomfortable) [inline consensus/service/src/byzantine_ledger/worker.rs:883]: Another subtle note is I don't know if from_contents_and_keypair signs %100 correctly. Comments on that in the method definition below.

File: transaction/core/src/blockchain/block_metadata.rs
Comment (iamalwaysuncomfortable) [inline transaction/core/src/blockchain/block_metadata.rs:99]: I'm unsure whether we should be using `try_sign_digest` or `try_sign` here. 

Notable
- `digest()` makes a hash of a `message` and I'm unsure whether the `digest` should be considered the `message` itself or whether the `digest` is a `pre-hash` of the message? Perhaps we should be signing a specific byte representation of `BlockMetaData` first? Any advice from @isis-mc or @garbageslam would be welcome here.
- I'm also unsure whether we should have any domain separators added to the digest first. I think we might want that?

**Review by iamalwaysuncomfortable — COMMENTED** [review]
Notes on the cryptography considerations - will review the rest later.

File: transaction/core/src/blockchain/quorum_set.rs
Comment (eranrund) [inline transaction/core/src/blockchain/quorum_set.rs:39]: Why two tags?

**Review by eranrund — COMMENTED** [review]
Nice work, overall looks great. Did this go through a k8 deployment?

**Review by iamalwaysuncomfortable — COMMENTED** [review]
This overall looks pretty great. My main questions are:
1. The crypto being used in signing the block metadata needs to be reviewed as I think we may want a domain separator and to double check the block_metadata digest is the correct message to hash
2a. Couple questions about the re-organization. I'd vote for leaving the transaction core crate methods where they are because there's work towards re-organizing that. 
2b. I'm okay with the verifier being moved into a types crate, but others might not be. I don't think our strategy for how to isolate types is formed, so it's a good proposal for maybe how to do that.
3. append_blocks could maybe avoid explicitly annotated lifetimes and we should sanity we're not running into any lifetime errors there vis-a-vis: https://github.com/pretzelhammer/rust-blog/blob/master/posts/common-rust-lifetime-misconceptions.md
4. We should create a tracking issue for all the todos in this PR

File: api/proto/blockchain.proto
Comment (jcape) [reply api/proto/blockchain.proto:76]: The idea is that the signature should be proving that a given quorum set and enclave were used when verifying a given block, and using the ID means you can prove that your "block 42" is the same as mine.

The idea is that you receive all of this at once, and check it as it comes in.

File: transaction/core/src/blockchain/quorum_set.rs
Comment (remoun) [reply transaction/core/src/blockchain/quorum_set.rs:20]: Making `GenericNodeId` include `prost::Message` proved problematic -- `mc_common::NodeID` and `mc_common::ResponderId` would also need to implement `Message`, and we don't currently have a `prost` dependency in `common`.

The other main issue was the impedance mismatch between how Prost handles `oneof` (hence `mc_transaction_core::QuorumSetMemberWrapper`) and how `scp::QuorumSet` is structured. I don't remember exactly what I got stuck on there, but I do remember it feeling gnarly, and giving up on that approach.

File: transaction/core/src/blockchain/block_data.rs
Comment (remoun) [reply transaction/core/src/blockchain/block_data.rs:7]: Oops, re-adding. It was a holdover from #1823, which removed `Message` so I could use `scp::QuorumSet`. Persisting to `LedgerDB` proved that `Message` is necessary.

File: transaction/core/src/blockchain/quorum_set.rs
Comment (remoun) [reply transaction/core/src/blockchain/quorum_set.rs:39]: `oneof` keys fields by tag number, and `QuorumSetMember` has two variants.

File: ledger/db/src/ledger_db.rs
Comment (remoun) [reply ledger/db/src/ledger_db.rs:153]: I had that initially, but passing `&None` felt clunky. I can switch back if `&None` is acceptable.

File: consensus/service/src/byzantine_ledger/worker.rs
Comment (remoun) [reply consensus/service/src/byzantine_ledger/worker.rs]: Added `log::crit!`

File: consensus/service/src/mint_tx_manager/mod.rs
Comment (remoun) [reply consensus/service/src/mint_tx_manager/mod.rs]: I've pruned many of them, and am trying to remove all of them with this PR. If any remain, I'll file a tracking issue.

File: attest/verifier/types/src/verification.rs
Comment (remoun) [reply attest/verifier/types/src/verification.rs]: I opted to split this crate per James' suggestion to minimize the additional footprint for these types. Suggestions welcome for where this crate might be better situated.

File: transaction/core/src/tx.rs
Comment (remoun) [reply transaction/core/src/tx.rs]: OK, reverted these changes.

File: transaction/core/src/blockchain/block_metadata.rs
Comment (remoun) [reply transaction/core/src/blockchain/block_metadata.rs]: Let's continue this discussion on https://github.com/mobilecoinfoundation/mobilecoin/pull/2058#discussion_r882283633

**Review by iamalwaysuncomfortable — APPROVED** [review]

File: consensus/service/src/byzantine_ledger/worker.rs
Comment (remoun) [reply consensus/service/src/byzantine_ledger/worker.rs]: #2104 tracks this

File: consensus/service/src/byzantine_ledger/worker.rs
Comment (remoun) [reply consensus/service/src/byzantine_ledger/worker.rs]: Agreed! Done.

**Comment by remoun** [comment]
This PR passed a test deploy.

File: ledger/sync/src/ledger_sync/ledger_sync_service.rs
Comment (remoun) [inline ledger/sync/src/ledger_sync/ledger_sync_service.rs:204]: Perhaps we shouldn't append the signature here, and possibly not even the metadata; see discussion at https://github.com/mobilecoinfoundation/mcips/pull/43/files#r898367411

File: blockchain/types/tests/digest.rs
Comment (remoun) [reply blockchain/types/tests/digest.rs:220]: Because I changed how the blocks are generated, and I couldn't find a way to make it compatible with the old blocks generations. So the options here are to change the bytes, or copy the old code that generated blocks here, which seems unwieldy.

File: consensus/service/src/byzantine_ledger/worker.rs
Comment (garbageslam) [reply consensus/service/src/byzantine_ledger/worker.rs:903]: The panic handler itself does a log crit, so it would make more sense to panic than to explicitly log crit and then panic

**Comment by remoun** [comment]
I pruned the test-utils refactoring and cleanup from this PR, for easier review. I'll send those out in another PR.

File: watcher/src/block_data_store.rs
Comment (remoun) [reply watcher/src/block_data_store.rs:46]: Good catch! I'll make it an optional 32 bytes.

File: ledger/db/src/ledger_db.rs
Comment (remoun) [reply ledger/db/src/ledger_db.rs:44]: Hmm, yeah. I lowered it to 19 and tests still pass -- I count 18 `Database`s under `ledger/db/`, and 1 for `MetadataStore`.

**Review by jcape — COMMENTED** [review]
The `hash()` method crept back in, otherwise looks good/will approve.

Thanks for trimming it down to size! :-)

**Comment by eranrund** [comment]
Is there ever a case where we would `append_block` with metadata but without a signature? My feeling is no, so perhaps we should not allow that (this will require changing a whole bunch of tests so it doesn't have to be in this PR)

**Review by eranrund — COMMENTED** [review]
LGTM pending @jcape's comments. I know we are currently having deployment issues, but it would be really nice if this is deployed and tested on an actual network before it gets merged.
It would also be good to test that LedgerDB/WatcherDB can successfully open and work on databases created prior to this merge. There's currently no automated test for this and it will be suboptimal if we find out that for whatever reason something here is incompatible with existing data.

File: ledger/db/src/ledger_trait.rs
Comment (remoun) [reply ledger/db/src/ledger_trait.rs:25]: I'm not sure; rustc is unhappy with the reference in an Option.

Without it I get an `error[E0637]: & without an explicit lifetime name cannot be used here` so I made all the references use the same lifetime.

Compiler output:
```
error[E0637]: `&` without an explicit lifetime name cannot be used here
  --> ledger/db/src/ledger_trait.rs:24:27
   |
24 |         signature: Option<&BlockSignature>,
   |                           ^ explicit lifetime name needed here

error[E0637]: `&` without an explicit lifetime name cannot be used here
  --> ledger/db/src/ledger_trait.rs:25:26
   |
25 |         metadata: Option<&BlockMetadata>,
   |                          ^ explicit lifetime name needed here

error[E0106]: missing lifetime specifier
  --> ledger/db/src/ledger_trait.rs:24:27
   |
24 |         signature: Option<&BlockSignature>,
   |                           ^ expected named lifetime parameter
   |
help: consider introducing a named lifetime parameter
   |
17 ~ #[automock]<'a>
18 | pub trait Ledger: Send {
19 |     /// Appends a block with contents, signature and metadata.
20 |     fn append_block<'b>(
21 |         &mut self,
22 |         block: &Block,
```

Comment (jcape) [reply ledger/db/src/ledger_trait.rs:25]: Blech, OK

Comment (garbageslam) [reply ledger/db/src/ledger_trait.rs:25]: oh so it's because of automock's code gen? that makes sense

**Review by eranrund — APPROVED** [review]
LGTM but would really prefer if this goes through k8 and a manual test of verifying existing Water/Ledger DBs work fine after this change, both when blocks with and without metadata appear.
Thank you!

**Comment by remoun** [comment]
> LGTM but would really prefer if this goes through k8 and a manual test of verifying existing Water/Ledger DBs work fine after this change, both when blocks with and without metadata appear. Thank you!

Agreed, I want to run both those tests before merging this. Currently blocked on broken deploys.

**Comment by remoun** [comment]
I'm pulling out the bit that makes block metadata required, since that's the tricky part of this PR that we want to test more.

**Comment by remoun** [comment]
> I'm pulling out the bit that makes block metadata required, since that's the tricky part of this PR that we want to test more.

Re-added since it's scoped to BlockVersion 3, and we're still at 0. We still need to address the catch-up mode, which I'll probably do in a follow-up PR.

**Comment by eranrund** [comment]
> > I'm pulling out the bit that makes block metadata required, since that's the tricky part of this PR that we want to test more.
> 
> Re-added since it's scoped to BlockVersion 3, and we're still at v0. We still need to address the catch-up mode, which I'll probably do in a follow-up PR.

What happens to mobilecoind instances that are deployed on test networks with block version 3 turned on?

**Comment by remoun** [comment]
> What happens to mobilecoind instances that are deployed on test networks with block version 3 turned on?

... They would fail. OK, reverting the required bit.

**Comment by remoun** [comment]
PTAL. I switched the `open_db` calls to `create_db` to work with existing DBs.

I'd started adding a LedgerDB migration but it ended up being just a `create_db` call, and found `create_db` works with existing `Database`s; I could add the migration if that is strongly preferred.

File: ledger/sync/src/ledger_sync/ledger_sync_service.rs
Comment (remoun) [reply ledger/sync/src/ledger_sync/ledger_sync_service.rs]: #2216 handles this

File: watcher/src/block_data_store.rs
Comment (eranrund) [inline watcher/src/block_data_store.rs:54]: I missed this in previous reviews, but I don't think this should be storing the metadata hash, similarly to how it doesn't store the signature hash.
The idea of storing hashes of the `Block` and `BlockContents` is that they are expected to be identical amongst all instances of a given block, but the signatures will vary for each source - or to phrase differently, for each node we "watch" we will be getting a list of `StoredBlockData` for the blocks produced by that node. For a given block index, the Block and BlockContents will very very likely be the same, so we avoid storing them multiple times in the database and instead store the hash here, and have a separate table to map hash -> actual contents.
We don't do that for the signature since that is expected to be unique, and in most cases the metadata will also be unique - so it should be stored here as `Option<BlockMetadata>`. This will likely simplify the changes here since there won't be a need for the new `BLOCK_METADATA_BY_HASH_DB_NAME` database.
Apologies for not catching it earlier.

**Review by eranrund — CHANGES_REQUESTED** [review]

File: watcher/src/block_data_store.rs
Comment (remoun) [reply watcher/src/block_data_store.rs:54]: Good catch!

> Apologies for not catching it earlier.

No worries; that's why I pared down this PR.

File: ledger/db/src/ledger_db.rs
Comment (garbageslam) [inline ledger/db/src/ledger_db.rs:38]: did we really go down from 27 to 19 here? or this number was just wrong before?

Comment (remoun) [reply ledger/db/src/ledger_db.rs:38]: The number was wrong before. I tried iteratively lowering it, 18 failed but 19 passed.

**Review by jcape — APPROVED** [review]

**Pull request merged**
